feature: added name constraints in certs with isCA enabled

tanujd11 opened the pull request:
<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

--> 

### Pull Request Motivation

Add name constraints feature in certificates with `isCA` enabled. Check https://github.com/cert-manager/cert-manager/issues/3655 for more details. The PR implements RFC: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10

Following is an example of the certificate using the nameConstraint:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: ca-cert-example
spec:
  secretName: example-ca-key-pair
  isCA: true
  issuerRef:
    name: selfsigned
    kind: ClusterIssuer
  commonName: "example1.com"
  dnsNames:
  - example1.com
  nameConstraints:
    critical: true
    permitted:
      dnsDomains: ["example1.com", "example2.com"]
      ipRanges: ["10.10.0.0/16"]
      emailAddress: ["example@example.org"]
    excluded:
      ipRanges: ["10.10.0.0/24"]
```

Verification: Checkout the nameConstraint section of the CA cert generated by above yaml.

```
openssl x509 -in cert.crt -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            51:81:28:14:0f:9b:fc:41:06:59:95:60:ff:2c:53:ed
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = example1.com
        Validity
            Not Before: Nov 23 06:11:27 2023 GMT
            Not After : Feb 21 06:11:27 2024 GMT
        Subject: CN = example1.com
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b7:fd:2c:bc:48:53:c2:f8:b8:2c:1d:54:62:ef:
                    04:d3:73:48:6d:24:2b:c9:37:cb:ba:7d:f3:b8:56:
                    f2:c7:59:26:70:a9:fd:7d:e2:34:3d:52:f9:ad:3d:
                    c9:54:17:1f:6e:92:56:cd:5f:4a:f2:fc:32:20:48:
                    94:b5:cf:30:8e:90:76:91:a5:34:bc:de:74:ab:8a:
                    a6:03:0b:41:74:f4:b8:c7:fd:93:19:e4:ed:d6:5f:
                    35:7f:2a:27:31:1c:9b:f0:92:05:c1:2f:19:73:de:
                    e1:b3:6b:6f:59:bd:89:5b:42:63:9c:82:6e:00:4c:
                    00:24:61:7f:4e:8d:95:0d:7b:b1:8b:2f:55:f8:d6:
                    a1:2e:0c:84:a4:64:38:a6:15:a5:1e:8d:8d:89:20:
                    55:52:15:ad:7b:ce:ac:b5:b8:f7:6b:18:24:54:0c:
                    63:fe:de:e4:1e:40:26:92:e8:e5:87:cf:23:0c:49:
                    f2:85:bb:94:ae:2f:4b:97:fb:97:ab:b5:b0:06:45:
                    51:69:95:d3:90:41:75:b1:4a:ed:6b:2d:fd:25:73:
                    51:83:8a:44:8c:f8:a0:20:dd:41:a4:d4:f5:4b:ef:
                    d8:6c:df:10:e6:51:67:15:c7:0d:ea:c3:d2:b1:64:
                    a5:57:04:00:5b:2e:91:31:97:ef:40:85:c5:9c:64:
                    74:67
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment, Certificate Sign
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Key Identifier:
                8B:00:98:38:52:1F:A1:03:83:12:B0:0A:D9:35:14:27:E2:FB:69:D4
            X509v3 Subject Alternative Name:
                DNS:example1.com
            X509v3 Name Constraints: critical
                Permitted:
                  DNS:example1.com
                  DNS:example2.com
                  IP:10.10.0.0/255.255.0.0
                  email:example@example.org
                Excluded:
                  IP:10.10.0.0/255.255.255.0
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        25:0f:c4:ea:e3:1b:29:c7:04:6a:ad:e1:78:bc:39:21:72:e3:
        a5:fc:4e:57:33:e1:92:c8:c3:41:b2:2b:22:ec:0f:2b:a8:3d:
        80:b3:3d:4b:c9:29:8e:fb:e9:0b:99:f5:2c:6c:02:b5:78:06:
        19:3b:55:f4:9e:1d:6a:b1:7d:3a:c8:07:47:50:68:d5:7d:c8:
        f6:83:45:6d:12:12:b5:15:ce:b9:e8:be:1c:81:ef:8a:01:08:
        66:e7:66:5c:02:fe:25:c7:f5:83:4e:9c:63:d2:3c:a3:eb:3a:
        77:8a:ef:8b:cc:fe:67:33:02:51:4f:81:d4:ff:8e:8b:e9:d0:
        7f:2a:f1:2e:51:5e:90:ea:ef:fd:62:54:9c:a0:82:1c:f5:ed:
        05:1a:a4:ee:aa:45:47:77:80:75:79:0c:04:d7:ee:4f:6a:1c:
        7e:f4:f7:83:d4:9a:e3:eb:28:13:88:fc:6c:70:df:c1:89:ab:
        76:b3:b7:e0:7f:63:53:bf:cc:92:c4:d2:4c:cb:ce:89:a3:dc:
        a3:09:13:14:88:aa:42:0a:87:df:86:3d:64:fe:36:cc:54:94:
        2e:51:77:2a:6c:98:ab:64:db:3b:cf:72:c9:9f:4f:48:bb:c3:
        f2:e2:cd:19:b7:f7:ea:29:0f:97:6c:65:92:ba:33:08:96:95:
        17:44:bc:50
-----BEGIN CERTIFICATE-----
MIIDdjCCAl6gAwIBAgIQUYEoFA+b/EEGWZVg/yxT7TANBgkqhkiG9w0BAQsFADAX
MRUwEwYDVQQDEwxleGFtcGxlMS5jb20wHhcNMjMxMTIzMDYxMTI3WhcNMjQwMjIx
MDYxMTI3WjAXMRUwEwYDVQQDEwxleGFtcGxlMS5jb20wggEiMA0GCSqGSIb3DQEB
AQUAA4IBDwAwggEKAoIBAQC3/Sy8SFPC+LgsHVRi7wTTc0htJCvJN8u6ffO4VvLH
WSZwqf194jQ9UvmtPclUFx9uklbNX0ry/DIgSJS1zzCOkHaRpTS83nSriqYDC0F0
9LjH/ZMZ5O3WXzV/KicxHJvwkgXBLxlz3uGza29ZvYlbQmOcgm4ATAAkYX9OjZUN
e7GLL1X41qEuDISkZDimFaUejY2JIFVSFa17zqy1uPdrGCRUDGP+3uQeQCaS6OWH
zyMMSfKFu5SuL0uX+5ertbAGRVFpldOQQXWxSu1rLf0lc1GDikSM+KAg3UGk1PVL
79hs3xDmUWcVxw3qw9KxZKVXBABbLpExl+9AhcWcZHRnAgMBAAGjgb0wgbowDgYD
VR0PAQH/BAQDAgKkMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFIsAmDhSH6ED
gxKwCtk1FCfi+2nUMBcGA1UdEQQQMA6CDGV4YW1wbGUxLmNvbTBfBgNVHR4BAf8E
VTBToEMwDoIMZXhhbXBsZTEuY29tMA6CDGV4YW1wbGUyLmNvbTAKhwgKCgAA//8A
ADAVgRNleGFtcGxlQGV4YW1wbGUub3JnoQwwCocICgoAAP///wAwDQYJKoZIhvcN
AQELBQADggEBACUPxOrjGynHBGqt4Xi8OSFy46X8Tlcz4ZLIw0GyKyLsDyuoPYCz
PUvJKY776QuZ9SxsArV4Bhk7VfSeHWqxfTrIB0dQaNV9yPaDRW0SErUVzrnovhyB
74oBCGbnZlwC/iXH9YNOnGPSPKPrOneK74vM/mczAlFPgdT/jovp0H8q8S5RXpDq
7/1iVJygghz17QUapO6qRUd3gHV5DATX7k9qHH7094PUmuPrKBOI/Gxw38GJq3az
t+B/Y1O/zJLE0kzLzomj3KMJExSIqkIKh9+GPWT+NsxUlC5RdypsmKtk2zvPcsmf
T0i7w/LizRm39+opD5dsZZK6MwiWlRdEvFA=
-----END CERTIFICATE-----
```
<!-- Explain the motivation behind this PR. If there's a related issue or PR, link to it here! -->

### Kind

<!--

Pick a kind which best describes your PR from the following list:

	<cleanup | bug | feature | documentation | design | flake>

If you're unsure which is best or if you're not sure what we mean by "kind",
just ignore this section and a maintainer will fill it in for you!
-->

feature

### Release Note

<!--

Should we mention this PR in release notes? If so, replace "NONE" with a line of text explaining what changed!

For more details, see: https://git.k8s.io/community/contributors/guide/release-notes.md

-->

```release-note
Users can now use nameConstraints in CA certificates. To know more details on name constraints check out RFC section https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10

```

jsoref commented on pkg/util/pki/nameconstraints.go:
```diff
@@ -0,0 +1,114 @@
+		ipNets = append(ipNets, &ipNet)
+	}
+	return ipNets
+}
```
could you include a newline at eof?

maelvls commented:
Hi! I'll turn on testing for you.

/ok-to-test

maelvls commented:
Hi! Great work on this, I'm really glad you were able to take a stab at this!! The API you propose looks good to me, no worries on this side. Regarding the choice of going with a feature flag, I imagine that the point of gating this feature is to refine the API surface? Or should we have to worry about the consequence of turning on this feature for everyone for some unexpected reason?

maelvls commented on internal/controller/feature/features.go:
```diff
@@ -98,6 +98,14 @@ const (
 	UseCertificateRequestBasicConstraints featuregate.Feature = "UseCertificateRequestBasicConstraints"
 
+	// Owner: @tanujd11
+	// Alpha: v1.13
```
Let's target 1.14 👍

```suggestion
	// Alpha: v1.14
```

tanujd11 commented:
Hi @maelvls I have added the API in the PR description. Also added validating name constraint step at issuer before signing the certificate. Please take a look. Thanks

maelvls commented:
For context for anyone reading this, this PR is in three folds:

- A new feature gate `UseCertificateRequestNameConstraints` introduced in alpha. This makes sense since we also gated the UseCertificateRequestBasicConstraints
- An addition to the Certificate API: the `nameConstraints` field is added. I think this change in not controversial.
- A pre-signing check specific to the CA issuer: when the CA issuer is about to sign a certificate, the CA issuer performs an ad-hoc check of the `nameConstraints` with respect to the DNS names, URIs, emails, and IP addresses.

**About the pre-signing check in the CA issuer:**


- The reason we want to add this pre-signing check is because Tanuj found out that it is easy to mistakenly to have the CA issuer issue a leaf certificate that doesn't comply with the name constraints. That's because the signed cert isn't validated after signing it in the CA issuer.
- For context, the name constraints are meant to be validated by the client during the TLS handshake. The client validates that the name constraints of the intermediate CAs and root CA match its subject DN and SANs.
- After looking at the complexity of the [Name Constraints section in the RFC](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10), I worry that this ad-hoc pre-signing check won't match the actual rules. The risk is minimal since it is the client's responsibility to check these constraints, not ours. The worst case scenario is that the pre-signing check is too tight, in which case we can always patch cert-manager.
- Go's x509 package has an internal func [`checkNameConstraints`](https://github.com/golang/go/blob/de5b418bea70aaf27de1f47e9b5813940d1e15a4/src/crypto/x509/verify.go#L501) that does this correctly, but it is much more involved and requires knowing about the whole chain, not just the (possibly) intermediate CA cert. The CA issuer doesn't have the information about the whole chain which prevents us from vendoring the Go lib's func.
- To conclude, although `validateNameConstraints` is imperfect, I am happy with keeping it as-is and not trying to re-implement the whole name constraints validation.

@SgtCoDFish Am I correct in thinking that the CA issuer can't validate its own issued certificates when the CA issuer is given an intermediate CA certificate and private key?

SgtCoDFish commented:
> @SgtCoDFish Am I correct in thinking that the CA issuer can't validate its own issued certificates when the CA issuer is given an intermediate CA certificate and private key?

My understanding is that the CA issuer can't guarantee that it can fully validate issued certificates because it can't guarantee it has access to its full chain.

Roots are less likely to have name constraints, so the most likely cause of failure here would be a chain `Root (R) -> Intermediate 1 (A) -> Intermediate 2 (B)` where both A and B have name constraints. In that case, if a CA issuer is configured with a secret containing just B (so the rest of the chain isn't available to it) then we won't be able to include the name constraints in A.

I'd consider that to be an edge case. I suspect there are unlikely to be multiple certs with name constraints in most chains, but there'll be some places which have that setup. Either case, I think it'd be fine to document that the CA issuer will only check for name constraints on the issuing cert. There's plenty of other stuff we don't check which could be present on the issuing cert (e.g. maxPathLen in basicConstraints)

TL;DR: I think we could skip the validation on the basis that we don't validate anything else currently. Validation will improve the UX for sure, but if it's hard to do we could remove it here. Later adding the validation would be good, but I don't think we need it for an initial implementation.

SgtCoDFish commented on internal/apis/certmanager/types_certificate.go:
```diff
@@ -531,3 +537,41 @@ type CertificateSecretTemplate struct {
+	// +optional
+	URIDomains []string
+	// if true then the name constraints are marked critical.
+	//
+	// +optional
+	Critical bool
```
suggestion: I don't think this should be under NameConstraintItem, I think it should be under NameConstraints

The API should look like this IMO:

```yaml
nameConstraints:
    critical: true
    permitted:
      dnsDomains: ["example1.com", "example2.com"]
      ipRanges: ["10.10.0.0/16"]
      emailAddress: ["example@example.org"]
    excluded:
      ipRanges: ["10.10.0.0/24"]
```

tanujd11 replied:
Currently critical is supported fir permittedDNSDomains only in go x509 lib. I had it in the NameConstraintItem if in future it supports different knob for excluded. I can move it to `nameConstraints` if it does not make sense though.

munnerz commented on deploy/crds/crd-certificates.yaml:
```diff
@@ -171,6 +171,64 @@ spec:
                 literalSubject:
                   description: "Requested X.509 certificate subject, represented using the LDAP \"String Representation of a Distinguished Name\" [1]. Important: the LDAP string format also specifies the order of the attributes in the subject, this is important when issuing certs for LDAP authentication. Example: `CN=foo,DC=corp,DC=example,DC=com` More info [1]: https://datatracker.ietf.org/doc/h ...
                   type: string
+                nameConstraints:
```
what do you think about naming this field something like `caNameConstraints`? Or, if we think there may be *more* CA-only options that we may add in future (not sure?) under some kind of `caOptions` structure/field?

inteon replied:
I think `caNameConstraints` could be confusing (do these nameConstraints apply to the CA?).
Also, the validating webhook will guide users to no create certificates that do not make sense (eg. setting `isCA: false` and `nameConstraints`)

tanujd11 replied:
nameConstraints as a setting is only valid for CA. Adding CA as prefix seems redundant. Also as @inteon suggested we have validation for nameConstraints to be used only if isCA is true. There is some similar code already used elsewhere where this is checked.

maelvls replied:
The Certificate spec has 19 fields currently. I'd tend to agree that the API is getting big:

```console
$ k explain certificate.spec | grep '^   [a-z]' | cut -d'<' -f1 | sort
additionalOutputFormats
commonName
dnsNames
duration
emailAddresses
encodeUsagesInRequest
ipAddresses
isCA
issuerRef
keystores
literalSubject
privateKey
renewBefore
revisionHistoryLimit
secretName
secretTemplate
subject
uris
usages
```

But as a new user or advanced user, I would probably not look at the API reference anyways. I would probably look for the X.509 field they want in their Certificate resource and do a Google each with it, for example:

- "missing usages certificate cert-manager"
- "encode usages certificate cert-manager"
- "java keystore certificate cert-manager"

On top of that, we know that the X.509 spec is "bound" and isn't growing anymore, we can be confident that we will add a few fields in the future but not that many.

As long as we specify that `nameConstraints` is only usable when `isCA` is set to true, I am OK with using the name `nameConstraints` since it matches the name in the RFC.

tanujd11 commented:
With all the arguments suggested to have the validateNameConstraints especially in this PR, I would like to propose that we should not have validateNameConstraints here at the moment. This could be part of some other PR taken separately. Also according to the RFC, the Cert validation is done by client which checks the nameConstraints in the cert chain. 

Please consider this comment to upvote/downvote for having/not having the validateNameConstraints. i.e

upvote if you think we should have validateNameConstraints as part of this PR.

@maelvls @inteon @munnerz @SgtCoDFish

SgtCoDFish commented on pkg/util/pki/nameconstraints.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// ConvertIPNeSliceToIPNetPointerSlice converts []net.IPNet to []*net.IPNet.
+func ConvertIPNeSliceToIPNetPointerSlice(ipNetPointerSlice []net.IPNet) []*net.IPNet {
```
It also has a typo; `ConvertIPNeSliceToIPNetPointerSlice` -> `ConvertIPNetSliceToIPNetPointerSlice`

tanujd11 commented:
Team I have addressed the review comments so far. Have also removed the `validateNameConstraints` function for now. PTAL. Thanks

inteon commented on internal/apis/certmanager/v1alpha2/types_certificate.go:
```diff
@@ -501,3 +507,41 @@ type CertificateAdditionalOutputFormat struct {
+	// if true then the name constraints are marked critical.
+	//
+	// +optional
+	Critical bool `json:"critical,omitempty"`
```
What should the default value be for Critical? @SgtCoDFish

tanujd11 replied:
I think the default should be false. Users need to configure this to true to make the constraints critical. The behaviour is similar in OpenSSL as in users would need to mention `critical` in name constraints for using this parameter

inteon commented on internal/apis/certmanager/validation/certificate.go:
```diff
@@ -149,6 +149,16 @@ func ValidateCertificateSpec(crt *internalcmapi.CertificateSpec, fldPath *field.
 	}
 
+	if crt.NameConstraints != nil {
+		if !crt.IsCA {
```
I think we should also add a feature gate check here.

tanujd11 replied:
The feature gate is at controller level and this validation is at webhook. Is there anything we can do to check the feature gate on controller from here. Introducing a feature gate on webhook help us nothing unless we make sure that both the controller and webhook have the same env vars. 

I hope I am making sense. Please correct me if not.

munnerz replied:
Both the webhook and controller must have feature gating - the webhook must have it to prevent people creating resources which won’t be valid due to the gate being disable. & yep, we expect users to configure the feature gate flag for both the webhook and the controller (similar to say, kube-apiserver and kube-controller-manager)

tanujd11 replied:
Thanks @munnerz for clarification. I have added the gate in webhook too. PTAL.

inteon commented on pkg/util/pki/nameconstraints.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// Adapted from x509.go
+func MarshalNameConstraints(nameConstraints *v1.NameConstraints) (pkix.Extension, error) {
```
I think it would be good to have a few unit tests for this function too.

inteon commented on deploy/crds/crd-certificates.yaml:
```diff
@@ -171,6 +171,61 @@ spec:
                   description: "Requested X.509 certificate subject, represented using the LDAP \"String Representation of a Distinguished Name\" [1]. Important: the LDAP string format also specifies the order of the attributes in the subject, this is important when issuing certs for LDAP authentication. Example: `CN=foo,DC=corp,DC=example,DC=com` More info [1]: https://datatracker.ietf.org/doc/h ...
                   type: string
+                nameConstraints:
+                  description: 'x.509 certificate NameConstraint extension which MUST be used only in a CA certificate. More Info: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10'
```
We probably want to make clear in this description that this feature can only be used if IsCA=true.
Additionally, this description should indication that this feature is only usable when the featuregate is set.

jsoref replied:
```suggestion
                  description: 'x.509 certificate NameConstraint extension which MUST NOT be used in a non-CA certificate. More Info: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10'
```

tanujd11 replied:
Done

inteon commented on make/e2e-setup.mk:
```diff
@@ -262,7 +262,7 @@ comma = ,
 # for "--set featureGates". That's why we have "\$(comma)".
-feature_gates_controller := $(subst $(space),\$(comma),$(filter AllAlpha=% AllBeta=% AdditionalCertificateOutputFormats=% ValidateCAA=% ExperimentalCertificateSigningRequestControllers=% ExperimentalGatewayAPISupport=% ServerSideApply=% LiteralCertificateSubject=% UseCertificateRequestBasicConstraints=% SecretsFilteredCaching=%, $(subst $(comma),$(space),$(FEATURE_GATES))))
+feature_gates_controller := $(subst $(space),\$(comma),$(filter AllAlpha=% AllBeta=% AdditionalCertificateOutputFormats=% ValidateCAA=% ExperimentalCertificateSigningRequestControllers=% ExperimentalGatewayAPISupport=% ServerSideApply=% LiteralCertificateSubject=% UseCertificateRequestBasicConstraints=% UseCertificateRequestNameConstraints=% SecretsFilteredCaching=%, $(subst $(comma),$(space),$(F ...
 feature_gates_webhook := $(subst $(space),\$(comma),$(filter AllAlpha=% AllBeta=% AdditionalCertificateOutputFormats=% LiteralCertificateSubject=%,   $(subst $(comma),$(space),$(FEATURE_GATES))))
```
The webhook also needs the `UseCertificateRequestNameConstraints` feature gate.

tanujd11 replied:
Thanks, Made the change.

inteon commented:
**Status update:**
I think this PR is almost ready, there seem to be only 2 remaining issues tasks:
- https://github.com/cert-manager/cert-manager/pull/6500#discussion_r1414073034
- https://github.com/cert-manager/cert-manager/pull/6500#discussion_r1415859081

tanujd11 commented:
Thanks @inteon for pointing this one out. Somehow I thought I fixed the typo but it was not the case. Have done it now, Also please let me know of your thoughts on the other [comment](https://github.com/cert-manager/cert-manager/pull/6500#discussion_r1415859081)

maelvls commented:
Hey! It seems like all the remarks from @inteon have been resolved. I'd be happy to have this merged once we have a documentation PR. Since this new feature applies to the next version of cert-manager (1.14), the changes need to be done to the `docs/` folder and **branched from the [release-next branch](https://github.com/cert-manager/website/tree/release-next)**. Let me know if you need help!

inteon reviewed (commented):
/approve
/lgtm
/hold in case someone else wants to review

maelvls reviewed (approved):
The documentation PR is ready: https://github.com/cert-manager/website/pull/1358. Thanks Tanuj!

/lgtm

inteon commented:
/unhold

inteon commented on pkg/util/pki/nameconstraints.go:
```diff
@@ -0,0 +1,117 @@
+
+// NameConstraints represents the NameConstraints extension.
+type NameConstraints struct {
+	PermittedDNSDomainsCritical bool        `asn1:"optional,explicit,tag:0"`
```
I missed this in the review, but this is not correct, this is not what is described in the x509 spec.
/cc @maelvls

inteon replied:
@tanujd11 could you create a new PR to fix this?

tanujd11 replied:
Hey @inteon , what exactly are you referring here?

inteon replied:
The marshalled NameConstraints extension in the CSR has to match the format defined in the RFC (see https://www.alvestrand.no/objectid/2.5.29.30.html).
The encoding you specified seems to be custom.
We can use the logic the x509 library uses: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/crypto/x509/x509.go;l=1189-1285;drc=82c713feb05da594567631972082af2fcba0ee4f

inteon commented:
NOTE: we found that we missed something in this PR and fixed the issue in https://github.com/cert-manager/cert-manager/pull/6542.

inteon commented on internal/controller/feature/features.go:
```diff
@@ -98,6 +98,14 @@ const (
+	// UseCertificateRequestNameConstraints will add Name Constraints section in the Extension Request of the Certificate Signing Request
+	// This feature will add NameConstraints section in CSR with CA field as true
+	// Github Issue: https://github.com/cert-manager/cert-manager/issues/3655
+	UseCertificateRequestNameConstraints featuregate.Feature = "UseCertificateRequestNameConstraints"
```
I think the feature gate is actually a bit different than the `UseCertificateRequestBasicConstraints` feature gate.
That feature gate only determines if the BasicConstraints have to be encoded in the CertificateRequest.
This feature gate will however also decide if this additional NameConstraints option in the Certificate object can be used at all.
